### ISSUE TITLE
Can the C compiler optimize SetUpvalue?

### ISSUE THREAD

**Issue by hugomg** [pallene-lang/pallene#453]
In the generated C code for the SetUpvalue, we are re-extracting the C Closure with `clCvalue` every time. Is this a problem? Maybe not: because there are no function calls in the generated code, I'm optimistic that the C compiler can optimize this. But we should check.

If the C compiler can optimize it, we should add a comment in the relevant  section of the coder.lua.

If the C compiler cannot optimize it, then we should figure out if we need to change the IR. For example, that idea of having a more complex recursion-aware version of NewClosure.

**Comment by hugomg** [comment]
For reference, here is the sort of code that we were talking about:

```c
    setobj(L, s2v(base + 1), &x3);
    setivalue(&(clCvalue(&x2)->upvalue[1]), x1);
    setobj(L, &(clCvalue(&x2)->upvalue[2]), &x3);
    setobj(L, s2v(base + 0), &x2);
    setivalue(&(clCvalue(&x2)->upvalue[1]), x1);
    setobj(L, &(clCvalue(&x2)->upvalue[2]), &x3);
    setobj(L, s2v(base + 0), &x2);
    setivalue(&(clCvalue(&x3)->upvalue[1]), x1);
    setobj(L, &(clCvalue(&x3)->upvalue[2]), &x2);
    setobj(L, s2v(base + 1), &x3);
    setivalue(&(clCvalue(&x3)->upvalue[1]), x1);
    setobj(L, &(clCvalue(&x3)->upvalue[2]), &x2);
    setobj(L, s2v(base + 1), &x3);
 ```
 
 I looked at the generated assembly and, as predicted, both gcc and clang can optimize the repeated `clCvalue()`. Provided that we compile with -O2, of course. With -O0 it generates repeated code as you would expect.

**Comment by hugomg** [comment]
Fixed in #467, even for -O0.
